### ISSUE TITLE
False positive from unnecessary_null_checks

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/linter#2608]
Create a file containing the following code:
```
  void f() {
    var s = getS();
    if (s == null) {
      return;
    }
    if (s.length > 2) {
      s = getS()!;
    }
    s.substring(0);
  }
  String? getS() => null;
```
and enable the `unnecessary_null_checks` lint. It will flag the null check, but removing the null check causes a diagnostic on the last line of `f` because the assignment to `s` prevents `s` from being promoted.

The lint should not flag uses of the null check operator when it changes promotion.

**Comment by pq** [comment]
@bwilkerson, @scheglov: is this tricky to fix?

**Comment by bwilkerson** [comment]
I don't know. What we need to know is that the promoted type of `s` before the assignment (on line 7) is not nullable, but I don't know whether the node for `s` gives us the type before or after the assignment.

**Comment by oprypin** [comment]
The vast majority of cases and the vast majority of value from this lint is right before passing something as a function argument. 

I think this lint shouldn't look at assignments local variables at all. I can't even come up with a *good* example of how this lint could possibly help out in local scope. Removing this false positive in this simple way will be worth it.

The way to detect local variables is probably just `scopeLookupResult` being `LocalVariableElement` || `ParameterElement`

**Comment by oprypin** [comment]
Linked a pull request accordingly
